From 9e77cdf01b73de313b360290ca0987410e04c082 Mon Sep 17 00:00:00 2001
From: Jacek Bukarewicz <j.bukarewicz@samsung.com>
Date: Wed, 3 Dec 2014 16:33:40 +0100
Subject: [PATCH] Perform Cynara runtime policy checks by default

By default clients are checked for http://tizen.org/privilege/internal/dbus
privilege for method calls, sending and receiving signals.
Services are supposed to override these defaults if certain resources
are going to be exposed to applications.
Method replies and errors are allowed by default.

Note that previously every message sent via session bus was allowed by
default. This has changed and session policy is now basically the same as
system configuration with 2 exceptions:
 - session bus only allows connections from the same user the session bus
   is running while system bus permits all users
 - requesting names is denied by default for system bus while session bus
   requires http://tizen.org/privilege/internal/dbus privilege

Change-Id: Ifb4a160bf6e0638404e0295a2e4fa3077efd881c
Signed-off-by: Jacek Bukarewicz <j.bukarewicz@samsung.com>
---
 bus/session.conf.in | 32 ++++++++++++++++++++++++++------
 bus/system.conf.in  | 29 ++++++++++++++++++++---------
 2 files changed, 46 insertions(+), 15 deletions(-)

diff --git a/bus/session.conf.in b/bus/session.conf.in
index 74d9d1f..fa5c232 100644
--- a/bus/session.conf.in
+++ b/bus/session.conf.in
@@ -17,12 +17,32 @@
   <standard_session_servicedirs />
 
   <policy context="default">
-    <!-- Allow everything to be sent -->
-    <allow send_destination="*" eavesdrop="true"/>
-    <!-- Allow everything to be received -->
-    <allow eavesdrop="true"/>
-    <!-- Allow anyone to own anything -->
-    <allow own="*"/>
+    <!-- By default clients require internal/dbus privilege to communicate
+         with D-Bus services and to claim name ownership. This is internal privilege that
+         is only accessible to trusted system services -->
+    <check own="*"                  privilege="http://tizen.org/privilege/internal/dbus" />
+    <check send_type="method_call"  privilege="http://tizen.org/privilege/internal/dbus" />
+    <check send_type="signal"       privilege="http://tizen.org/privilege/internal/dbus" />
+    <check receive_type="signal"    privilege="http://tizen.org/privilege/internal/dbus" />
+
+    <!-- Reply messages (method returns, errors) are allowed
+         by default -->
+    <allow send_requested_reply="true" send_type="method_return"/>
+    <allow send_requested_reply="true" send_type="error"/>
+
+    <!-- All messages but signals may be received by default -->
+    <allow receive_type="method_call"/>
+    <allow receive_type="method_return"/>
+    <allow receive_type="error"/>
+
+    <!-- Allow anyone to talk to the message bus -->
+    <allow send_destination="org.freedesktop.DBus"/>
+    <allow receive_sender="org.freedesktop.DBus"/>
+
+    <!-- But disallow some specific bus services -->
+    <deny send_destination="org.freedesktop.DBus"
+          send_interface="org.freedesktop.DBus"
+          send_member="UpdateActivationEnvironment"/>
   </policy>
 
   <!-- Config files are placed here that among other things, 
diff --git a/bus/system.conf.in b/bus/system.conf.in
index 92f4cc4..f3fd974 100644
--- a/bus/system.conf.in
+++ b/bus/system.conf.in
@@ -46,25 +46,36 @@
     <allow user="*"/>
 
     <!-- Holes must be punched in service configuration files for
-         name ownership and sending method calls -->
-    <deny own="*"/>
-    <deny send_type="method_call"/>
+         name ownership -->
+    <deny own="*" />
 
-    <!-- Signals and reply messages (method returns, errors) are allowed
+    <!-- By default clients require internal/dbus privilege to communicate
+         with D-Bus services. This is internal privilege that is only accessible
+         to trusted system services -->
+    <check send_type="method_call"  privilege="http://tizen.org/privilege/internal/dbus" />
+    <check send_type="signal"       privilege="http://tizen.org/privilege/internal/dbus" />
+    <check receive_type="signal"    privilege="http://tizen.org/privilege/internal/dbus" />
+
+    <!-- Reply messages (method returns, errors) are allowed
          by default -->
-    <allow send_type="signal"/>
     <allow send_requested_reply="true" send_type="method_return"/>
     <allow send_requested_reply="true" send_type="error"/>
 
-    <!-- All messages may be received by default -->
+    <!-- All messages but signals may be received by default -->
     <allow receive_type="method_call"/>
     <allow receive_type="method_return"/>
     <allow receive_type="error"/>
-    <allow receive_type="signal"/>
 
-    <!-- Allow anyone to talk to the message bus -->
+    <!-- If there is a need specific bus services could be protected by Cynara as well.
+         However, this can lead to deadlock during the boot process when such check is made and
+         Cynara is not yet activated (systemd calls protected method synchronously,
+         dbus daemon tries to consult Cynara, Cynara waits for systemd activation).
+         Therefore it is advised to allow root processes to use bus services.
+         Currently anyone is allowed to talk to the message bus -->
     <allow send_destination="org.freedesktop.DBus"/>
-    <!-- But disallow some specific bus services -->
+    <allow receive_sender="org.freedesktop.DBus"/>
+
+    <!-- Disallow some specific bus services -->
     <deny send_destination="org.freedesktop.DBus"
           send_interface="org.freedesktop.DBus"
           send_member="UpdateActivationEnvironment"/>
-- 
2.1.4

